test/src/lib/fs-utils.spec.ts

```diff
@@ -0,0 +1,157 @@
+import * as path from 'path';
+import { promises as fs } from 'fs';
+import { spy, SinonSpy } from 'sinon';
+import mock = require('mock-fs');
+
+import * as fsUtils from '../../../src/lib/fs-utils';
+import { rootMountPoint } from '../../../src/lib/constants';
+
+describe('lib/fs-utils', () => {
+	const testFileName1 = 'file.1';
+	const testFileName2 = 'file.2';
+	const testFile1 = path.join(rootMountPoint, testFileName1);
+	const testFile2 = path.join(rootMountPoint, testFileName2);
+
+	const mockFs = () => {
+		mock({
+			[testFile1]: 'foo',
+			[testFile2]: 'bar',
+		});
+	};
+
+	const unmockFs = () => {
+		mock.restore();
+	};
+
+	describe('writeAndSyncFile', () => {
+		before(mockFs);
+		after(unmockFs);
+
+		it('should write and sync string data', async () => {
+			await fsUtils.writeAndSyncFile(testFile1, 'foo bar');
+			expect(await fs.readFile(testFile1, 'utf-8')).to.equal('foo bar');
+		});
+
+		it('should write and sync buffers', async () => {
+			await fsUtils.writeAndSyncFile(testFile1, Buffer.from('bar foo'));
+			expect(await fs.readFile(testFile1, 'utf-8')).to.equal('bar foo');
+		});
+	});
+
+	describe('writeFileAtomic', () => {
+		before(() => {
+			spy(fs, 'rename');
+			mockFs();
+		});
+
+		after(() => {
+			(fs.rename as SinonSpy).restore();
+			unmockFs();
+		});
+
+		it('should write string data atomically', async () => {
+			await fsUtils.writeFileAtomic(testFile1, 'foo baz');
+			expect(await fs.readFile(testFile1, 'utf-8')).to.equal('foo baz');
+			expect(fs.rename).to.have.been.calledWith(`${testFile1}.new`, testFile1);
+		});
+
+		it('should write buffer data atomically', async () => {
+			await fsUtils.writeFileAtomic(testFile1, 'baz foo');
+			expect(await fs.readFile(testFile1, 'utf-8')).to.equal('baz foo');
+			expect(fs.rename).to.have.been.calledWith(`${testFile1}.new`, testFile1);
+		});
+	});
+
+	describe('safeRename', () => {
+		beforeEach(mockFs);
+		afterEach(unmockFs);
+
+		it('should rename a file', async () => {
+			await fsUtils.safeRename(testFile1, testFile1 + 'rename');
+			const dirContents = await fs.readdir(rootMountPoint);
+			expect(dirContents).to.have.length(2);
+			expect(dirContents).to.not.include(testFileName1);
+			expect(dirContents).to.include(testFileName1 + 'rename');
+		});
+
+		it('should replace an existing file', async () => {
+			await fsUtils.safeRename(testFile1, testFile2);
+			const dirContents = await fs.readdir(rootMountPoint);
+			expect(dirContents).to.have.length(1);
+			expect(dirContents).to.include(testFileName2);
+			expect(dirContents).to.not.include(testFileName1);
+		});
+	});
+
+	/**
+	 * TODO: Un-skip this test after all fs tests that write to a test file system use
+	 * mock-fs instead. Hypothesis: exists isn't handling the relative directory it's
+	 * being passed well. When all unit tests use mock-fs, we can set process.env.ROOT_MOUNTPOINT
+	 * to `/mnt/root` so we can have an absolute path in all these tests.
+	 */
+	describe.skip('existsAsync', () => {
```

Comment:
Ignore what I mentioned before about this test suite finally working with the `npm run clean` script, it still works with `npm run test:fast` but not `npm run test`. After exhausting all other possibilities, the most likely hypothesis for me is that this is because of the directory level differences between build/ and test/, even though theoretically this feels like this is not the case. In another PR this week I want to refactor all fs-using tests to use mock-fs, then set process.env.ROOT_MOUNTPOINT in `00-init.spec.ts` to an absolute path instead of a relative one - I think this will solve the problem.